Matlab/Matlab.sublime-syntax

```diff
@@ -48,477 +76,792 @@ contexts:
+    - match: \b(exist)(?:\s+(\w+)(?:\s+(builtin|class|dir|file|var))?)?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: variable.other.matlab
+        3: storage.type.matlab
+    # https://www.mathworks.com/help/matlab/ref/exit.html
+    # https://www.mathworks.com/help/matlab/ref/quit.html
+    - match: \b(exit|quit)(?:\s+(cancel|force))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/format.html
+    - match: \b(format)(?:\s+(short|long|shortE|longE|shortG|longG|shortEng|longEng|bank|hex|rational|compact|loose))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/grid.html
+    - match: \b(grid)(?:\s+(on|off|minor))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/hidden.html
+    - match: \b(hidden)(?:\s+(on|off))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/hold.html
+    - match: \b(hold)(?:\s+(on|off|all))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/import.html
+    - match: \b(import)\b(?:\s+((?:\w+\.)*(?:\w+|\*)?))?
+      captures:
+        1: keyword.other.import.matlab
+        2: meta.path.import.matlab
+    # https://www.mathworks.com/help/matlab/ref/material.html
+    - match: \b(material)(?:\s+(shiny|dull|metal|default))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/mex.html
+    - match: \bmex\b(?!\.)
+      scope: keyword.other.command.matlab
+    # https://www.mathworks.com/help/matlab/ref/more.html
+    - match: \b(more)(?:\s+(on|off))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/opengl.html
+    - match: \b(opengl)(?:\s+(info|software|hardware|hardwarebasic))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: support.constant.matlab
+    # https://www.mathworks.com/help/matlab/ref/who.html
+    # https://www.mathworks.com/help/matlab/ref/whos.html
+    - match: \b(who|whos)(?:\s+(global))?\b
+      captures:
+        1: keyword.other.command.matlab
+        2: storage.modifier.matlab
+
+  # Parentheses are usually omitted when functions are used without parameters.
+  # This command syntax is also possible for functions with string parameters.
+  # Only the function names are matched here to allow both command syntax and function syntax form.
+  # https://www.mathworks.com/help/matlab/matlab_prog/command-vs-function-syntax.html
+  builtin-functions:
```

Comment:
Maybe combining all the different lists of builtin functions and classes in the `variables` section may make sense?

To illustrate what I have in mind:

https://github.com/sublimehq/Packages/blob/4b66388fd5b206357769a9650c163f1a6958cea6/CSS/CSS.sublime-syntax#L55-L72

and than just use:

https://github.com/sublimehq/Packages/blob/4b66388fd5b206357769a9650c163f1a6958cea6/CSS/CSS.sublime-syntax#L2536-L2549

Another example may be the different HTML tag types in CSS.

Reply:
I'm not sure whether it makes sense here to have all function names defined as a variable, because they are used only once in this syntax.

The classification represents the function lists from the Matlab docs (left side menu at https://www.mathworks.com/help/matlab/referencelist.html?type=function&listtype=cat&category=index&blocktype=all&capability=&s_tid=CRUX_lftnav).
I hope that I've removed all duplicates and "non-functions". To merge all lists would probably make it easier to find duplicates.

Reply:
The variables approach in general has following motivations:

1. avoid those endless lines and replace them by hard wrapped code blocks of fixed width
2. variables are not indented and therefore can use most available space to achieve (1.)
3. variables may help inerhited syntaxes a bit to replace or extend certain kinds of variables. It would require to group commands from different libraries in dedicated variables though. This is what the "units" example should illustrate. The html tags variables in CSS may be a better one though.

All of those are experiences I've made. They are of rather low priority for Matlab as it is quite unlikely it being extended or inherited.

Reply:
Okay, I've merged most of the function lists and moved them to the variables section.

Your third point might actually be quite useful even for Matlab; there is the Octave software, which aims to be a free and open source replacement for Matlab, with mostly compatible syntax. Not all Matlab functions are implemented in Octave though, and there are a few other extension. For example in the "Indentation Rules.tmPreferences" file in this Matlab package, there are patterns for `unwind_protect` and for `end\w*` (instead of just `end`), which are Octave-specific and have no meaning in Matlab. Perhaps they should better be removed here. Some snippets seem to be targeting Octave as well. There is also a [OcataveSupport](https://packagecontrol.io/packages/OctaveSupport) package on Package Control with its own syntax, but it seems to be just a copy of this Matlab syntax from an earlier state with adjusted header.

 And to the indentation patterns file for Matlab should be added `parfor`, `spmd` and `arguments`, but maybe that's another PR.

Reply:
I guess the solution depends on two factors:

1. How easy is it to select correct syntax for an end user? As Octave makes use of the same file extensions it might require active setup or interaction to choose correct dialect.
2. Whether or not a dialect such as Octave is interpreted as new syntax and/or is accepted to be added to default packages.

Basically a syntax should keep focused and precise. On the other hand end users might expect "Octave" support if those syntaxes are mostly compatible. Python for instance also supports both 2 and 3 which have slightly different lists of builtins.

If there are no unsolvable conflicts and differences are builtin functions only, it might be the simples solution to just add those additional Octave functions to Matlab.

An alternative might be to add an Octave syntax as inherited one, if sublimehq is interested. I have an impression of trying to keep syntaxes which "inherite each other" together might be the most robust solution.

Reply:
I have only used Matlab and not Octave, but according to the [Octave Wiki](https://wiki.octave.org/wiki/index.php?title=Differences_between_Octave_and_Matlab#What_features_are_unique_to_Octave.3F) there seem to be quite some differences which would need changes or additions to this syntax. These are not only built-in functions, but also various new operators, a few more keywords and block-constructs, and another character for comments. And I think none of the Octave-specific syntax is currently recognized by this Matlab syntax (except for the .tmPreferences file and the snippets), so it's probably better to leave it to the Octave package on Package Control.